### ISSUE TITLE
Best Practices to avoid changing permissions in a container?

### ISSUE THREAD

**Issue by bnmnetp** [RunestoneInteractive/RunestoneServer#1258]
Hi @vsoch 

So, right now in our entrypoint.sh there is a place where I change the ownership on everything to www-data.  Because I work on a mac, this fooled me into think all was fine.  But I've recently switched to using a cloud server for most development, and under Linux I don't own my files anywhere.  My limited searching has revealed that my approach is not best practice, but I'm wondering if you have a better (yet still simple) solution to this?

**Comment by vsoch** [comment]
Can you give me more detail? I don't understand what you mean by "I don't own my files anywhere." Are you saying that www-data is on the host or in the container? And where are you issuing the command from? Do you mean that you are on a server where you don't have root (why, and how can you use docker?) If you are issuing the command from inside a Docker container, you should be root always. If you are issuing the command from outside Docker, why are you trying to change the permissions on a folder on the host (and should the host allow you to?)

**Comment by bnmnetp** [comment]
On lines 96 and following in entrypoint.sh I have the commands

```
info "Updating file ownership"
chown -R www-data /srv/web2py
mkdir -p /run/uwsgi
chown -R www-data /run/uwsgi
```

In the container this is great because both nginx and uwsgi run as user www-data.

However, what I was surprised about was that this changed the ownership of the files in the shared volume on the host side as well.   This did not happen on the mac, I retained ownership of the shared files on the host side even as they were correctly owned by www-data in the container.

This is causing some colleagues a lot of problems because they can no longer do a git pull in their development environment once we have started the container.  I got around this by changing group permissions on linux.  I had the impression there were better solutions...

**Comment by vsoch** [comment]
lol  The fire alarm went off! I’ll be able to respond when I can get back inside :-)

**Comment by vsoch** [comment]
I can’t speak for Mac, but if you have a file or folder bound and you change permissions, it will propagate to the host. The only solution is not to bind the folder.

**Comment by bnmnetp** [comment]
OK, my guess is that on the mac its different than linux because it needs to actually use a VM whereas linux uses other tricks for virtualization.

**Comment by vsoch** [comment]
I’m looking here https://github.com/boot2docker/boot2docker/issues/587 and it seems like the permissions aren’t handled properly, or the Mac version creates some different volume on the host.

**Comment by vsoch** [comment]
Linux doesn’t need a VM to run containers, it’s run connecting to kernel on the host. I’d guess Mac does need a VM with a Linux kernel, but I’m not sure.

**Comment by vsoch** [comment]
I don't know if there are any tricks around this, but file ownership (I think) tends to match UID and GID. So if you are able to create ids in the container that match some set on the host (and likely a group makes sense) that's the only thing I can think of that might help.

I've actually never run anything Docker in a shared space, usually the instances I bring up are just me :)

**Comment by bjones1** [comment]
On my server, I run Apache and made all files to be in the `www-data` group, and that plays well with both Apache, web2py, and git. Would that work for nginx in Docker, since changing ownership to the `www-data` user causes problems?

**Comment by bnmnetp** [comment]
@bjones1 - what I have done in my linux environment is to make sure that both www-data and my username are in a common group.  Then I made sure all of my dev files are group writeable.  Problem solved.

**Comment by bjones1** [comment]
This was fixed in 2281031366eafca63d385168ad0b01b1721bc2b4. @bnmnetp, you can close this.
